### PR TITLE
Experiment/visualize perturbations

### PR DESCRIPTION
**Pull request by ihowell**
This is a WIP PR. We are only merging now because work includes plotting images in a grid.

### CODE REVIEW

File: pyroclast/adversarial/adversarial.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import matplotlib.pyplot as plt
+
+from pyroclast.common.tf_util import load_model
+from pyroclast.common.plot import plot_images
+from pyroclast.common.adversarial import fast_gradient_method
+from pyroclast.features.features import build_savable_objects
+
+
+def visualize_perturbation(data_dict, seed, output_dir, debug, module_name,
+                           model_name, norm, data_index, epsilon, **kwargs):
```
Comment (DrKwint) [inline pyroclast/adversarial/adversarial.py]: Might doc this, especially the assumptions made about the module being imported.

Comment (DrKwint) [reply pyroclast/adversarial/adversarial.py]: Also, the fact that this is attacking the usefulness of features

Comment (DrKwint) [reply pyroclast/adversarial/adversarial.py]: What does `data_index` do?

File: pyroclast/adversarial/adversarial.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        y = batch_data['label'][0]
+        break
+    x = tf.cast(x, dtype=tf.float32) / 255.
+
+    x = tf.reshape(x, [1] + x.shape)
+    y = tf.reshape(y, [1] + y.shape)
+
+    num_classes = 10
+    feature_idx = 0
+    class_idx = 0
```
Comment (DrKwint) [inline pyroclast/adversarial/adversarial.py]: Should these be hard coded?

File: pyroclast/adversarial/adversarial.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+                       output_dir=output_dir,
+                       **kwargs)
+    for batch_data in data_dict['train']:
+        x = batch_data['image'][0]
+        y = batch_data['label'][0]
+        break
+    x = tf.cast(x, dtype=tf.float32) / 255.
+
+    x = tf.reshape(x, [1] + x.shape)
+    y = tf.reshape(y, [1] + y.shape)
```
Comment (DrKwint) [inline pyroclast/adversarial/adversarial.py]: Prefer to use `tf.expand_dims` here. It's more robust and self-documenting.

File: pyroclast/adversarial/adversarial.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+    num_classes = 10
+    feature_idx = 0
+    class_idx = 0
+
+    def get_one_hot(x, num_classes):
+        return tf.cast(tf.one_hot(x, num_classes, on_value=1, off_value=-1),
+                       tf.float32)
+
+    labels = get_one_hot(y, num_classes)
```
Comment (DrKwint) [inline pyroclast/adversarial/adversarial.py]: re-name to something like `one_hot_labels`?

File: pyroclast/adversarial/adversarial.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+    x_perturbation = fast_gradient_method(forward_fn, x, epsilon, norm)
+
+    perturbed_x = x + x_perturbation
+
+    plot_images([[x, x_perturbation, perturbed_x] for _ in range(3)],
+                row_labels=['Images 1', 'Images 2', 'Images 3'],
+                col_labels=['Original', 'Perturbation', 'Perturbed'])
+
+    plt.show()
```
Comment (DrKwint) [inline pyroclast/adversarial/adversarial.py]: Should this show, save, or do both? Should these be controllable?

File: pyroclast/common/tf_util.py [hunk 6]
```diff
@@ -283,3 +283,30 @@ def finalize(self):
             self.sum_first)) * tf.math.sqrt(self.n * self.sum_sq_second -
                                             tf.math.square(self.sum_second))
         return numerator / (denominator + 1e-12)
+
+
+def load_model(module,
+               model_save_name,
+               data_dict,
+               output_dir=__file__ + '/../../../',
```
Comment (DrKwint) [inline pyroclast/common/tf_util.py]: This seems very specific...

File: pyroclast/common/tf_util.py [hunk 7]
```diff
@@ -283,3 +283,30 @@ def finalize(self):
...
+               conv_stack_name='ross_net',
+               learning_rate=1e-4):
+    # model_dir = os.abspath(os.path.normpath(output_dir + model_save_name))
+    # print('model_dir', model_dir)
```
Comment (DrKwint) [inline pyroclast/common/tf_util.py]: comments should be stricken

File: pyroclast/common/tf_util.py [hunk 8]
```diff
@@ -283,3 +283,30 @@ def finalize(self):
...
+    build_savable_objects_func = getattr(module, 'build_savable_objects')
+    if build_savable_objects_func is None:
+        raise Exception(
+            "Given module does not have a `build_savable_objects` method.")
+    objects = build_savable_objects_func(conv_stack_name, data_dict,
+                                         learning_rate, output_dir,
+                                         model_save_name)
+
+    print('objects', objects)
+    print('latest', objects['ckpt_manager'].latest_checkpoint)
```
Comment (DrKwint) [inline pyroclast/common/tf_util.py]: prints should be stricken

File: pyroclast/adversarial/adversarial.py (hunk 2 above)
Comment (ihowell) [reply pyroclast/adversarial/adversarial.py]: Obviously not. All of this is going to change.

File: pyroclast/adversarial/adversarial.py (hunk 1 above)
Comment (ihowell) [reply pyroclast/adversarial/adversarial.py]: Going to take the index of data from the training data. If larger, test data. This is probably going to change. This file shouldn't be in question, I just needed a PR for the plot.

File: pyroclast/adversarial/adversarial.py (hunk 3 above)
Comment (ihowell) [reply pyroclast/adversarial/adversarial.py]: Agreed. Will change, just not on this PR. Need to get plot data in.

File: pyroclast/adversarial/adversarial.py (hunk 4 above)
Comment (ihowell) [reply pyroclast/adversarial/adversarial.py]: Sounds good.

File: pyroclast/adversarial/adversarial.py (hunk 5 above)
Comment (ihowell) [reply pyroclast/adversarial/adversarial.py]: This is just an experiment at the moment, so I'm just looking over them by hand. Unsure of what we are going to do in the end.

File: pyroclast/common/tf_util.py (hunk 6 above)
Comment (ihowell) [reply pyroclast/common/tf_util.py]: It sort of is. __file__ is very specific so we have to adjust to it. Currently, it points to the project root directory.

**Comment by ihowell** [comment]
I agree that `pyroclast.adversarial.adversarial` should change, and it will, drastically. Just made this PR to get in some plotting code.

**Pull request merged**